MobiFlight/MidiBoard/MidiBoardManager.cs

```diff
@@ -0,0 +1,273 @@
+        }
+
+        private List<string> CollectErrorsInOutputDefinition(MidiBoardDefinition def)
+        {
+            List<string> errorMessages = new List<string>();
+            if (def.Outputs != null)
+            {
+                foreach (var item in def.Outputs)
+                {
+                    if (item.LabelIds.Length != item.MessageIds.Length)
+                        errorMessages.Add($"{item.Label}: Number of LabelIds unequal to number of MessageIds");
+                    if (!item.Label.Contains("%"))
+                        errorMessages.Add($"{item.Label}: Label does not contain % placeholder.");
+                    if (!string.IsNullOrEmpty(item.RelatedInputLabel))
+                    {
+                        if (item.RelatedIds.Length != item.MessageIds.Length)
+                            errorMessages.Add($"{item.Label}: Number of RelatedIds unequal to number of MessageIds");
+                        if (!item.RelatedInputLabel.Contains("%"))
+                            errorMessages.Add($"{item.Label}: RelatedInputLabel does not contain % placeholder.");
+                    }
+                }
+            }
+            return errorMessages;
+        }
+
+
+        private bool IsErrorInDefinition(MidiBoardDefinition def)
+        {
+            List<string> errorMessages = new List<string>();
+            errorMessages.AddRange(CollectErrorsInInputDefinition(def));
+            errorMessages.AddRange(CollectErrorsInOutputDefinition(def));
+            foreach (var error in errorMessages)
+            {
+                Log.Instance.log($"[{def.InstanceName}]: {error}", LogSeverity.Error);
+            }
+            return errorMessages.Count > 0;            
+        }
+
+
+        private void LoadDefinitions()
+        {
+            foreach (var definitionFile in Directory.GetFiles("MidiBoards", "*.midiboard.json"))
+            {
+                try
+                {
+                    var midiBoardDef = JsonConvert.DeserializeObject<MidiBoardDefinition>(File.ReadAllText(definitionFile));
+                    if (!IsErrorInDefinition(midiBoardDef))
+                    {
+                        Definitions.Add(midiBoardDef.InstanceName, midiBoardDef);
+                        Log.Instance.log($"Loaded midiBoard definition for {midiBoardDef.InstanceName}", LogSeverity.Info);
+                    }                    
+                }
+                catch (Exception ex)
+                {
+                    Log.Instance.log($"Unable to load {definitionFile}: {ex.Message}", LogSeverity.Error);
+                }
+            }         
+        }
+
+        public bool MidiBoardsConnected()
+        {
+            return MidiBoards.Count > 0;
+        }
+      
+        public void Startup()
+        {
+            ProcessTimer.Start();
+        }
+
+        private void ProcessTimer_Tick(object sender, EventArgs e)
+        {    
+            if (BoardsToBeRemoved.Count > 0)
```

Comment:
I don't think this if is needed. You can just run the `foreach` and if there are no boards to remove it'll do nothing. There's no issue calling `BoardsToBeRemoved.Clear()` every tick either since it'll noop if there are no boards to be removed.

Reply:
Yeah, I can remove the check. The same check is done within the .Clear() Method in the framework.